### DISCUSSION TITLE
Abstract Patterns in SCH v/s Pure

### DISCUSSION THREAD

**Discussion by pradyumanaggarwal** [phax/ph-schematron#146] [General]
I was using SchematronResourcePure implementation, and trying out abstract pattern. But, same abstract pattern works in pure implementation but not in SCH. Getting few errors but if i tried to resolve those errors, then they dissolves the purpose of abstract patterns. 

Schematron file : 
```
<?xml version="1.0" encoding="UTF-8"?>
<sch:schema xmlns="http://purl.oclc.org/dsdl/schematron" xmlns:sch="http://purl.oclc.org/dsdl/schematron"
  xmlns:sqf="urn:anything"
  queryBinding="xslt2" schemaVersion="ISO19757-3">  <sch:title>Schematron 1</sch:title>
  <sch:pattern abstract="true" id="LimitNoOfWords">
   <sch:rule context="$parentElement">
     <sch:let name="words" value="string-length(.)"/>
     <sch:assert test="$words &gt; $maxWords">
    You have <sch:value-of select="$words"/> word(s).
     </sch:assert>
     <sch:assert test="$words &lt; $minWords">
    You have <sch:value-of select="$words"/> word(s).
     </sch:assert>
   </sch:rule>
	</sch:pattern>
  <sch:pattern is-a="LimitNoOfWords">
   <sch:param name="parentElement" value="title"/>
    	<param name="minWords" value="1"/>
    	 <param name="maxWords" value="8"/>
  </sch:pattern>
  
</sch:schema>
```

SCH Implementation : 
```
final ISchematronResource schemaResource = SchematronResourceSCH.fromFile(this.schematronFile);
SchematronOutputType svrl = schemaResource.applySchematronValidationToSVRL(doc, null);
```
where doc is xml docuement.

Pure Implementation : 
```
ISchematronResource schemaResource = SchematronResourcePure.fromInputStream(streamSchematron);
aResSchematron.applySchematronValidationToSVRL(dom);
```
where dom is xml node.

Anyone has any idea  ? @phax

**Comment by phax** [comment]
~Eventally, because the last 2 params use `<param` instead of `<sch:param` ???~
Sorry, just saw that you have both namespaces on top

I found the issue: the `pattern` that uses `is-a` also needs to have an `id` attribute. Then it works.
@pradyumanaggarwal

**Comment by pradyumanaggarwal** [comment]
When using SchematronResourceSCH, can we write XPATHS in context ? 

For ex : 

XML : 
```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE topic PUBLIC "-//OASIS//DTD DITA Topic//EN" "technicalContent/dtd/topic.dtd">
<topic id="GUID-84fa6d91-8506-4d06-9595-ebabc6f8b46f-enHYPHENUS" xml:lang="en-US">
  <title>Sample</title>
  <shortdesc>This short description is too short. Or too long? The recommendation is, that the short description should have around twenty words.</shortdesc>
  <body>
    <p>This text is using the &quot;straight quotes&quot; which should be reported.</p>
    <p>This text is using the «straight quotes» which should be reported.</p>
    <p>This paragraph contains the word Content.</p>
  </body>
</topic>
```

Here, i won't get any output if i use this schematron file : 

```
<?xml version="1.0" encoding="UTF-8"?>
<sch:schema xmlns:sch="http://purl.oclc.org/dsdl/schematron" queryBinding="xslt2">
   <sch:title>Schematron Stefan</sch:title>
   <sch:pattern id="quotation-marks-en">
      <sch:rule context="//text()[ancestor::*[@xml:lang = 'en-US']]">
         <sch:assert role="info" subject="text()" test="not(matches(., '«|»|„'))">Please use the correct typographic quotes for the English language (“” and ‘’).</sch:assert>
         <sch:assert role="info" subject="text()" test="matches(., '«|»|„')">Please use the correct typographic quotes for the English language (“” and ‘’).</sch:assert>
      </sch:rule>
   </sch:pattern>
</sch:schema>
```


But i will get validation errors if i use `context="body"` .
@phax

**Reply by phax** [reply]
I think this is an XPath issue. I added another test in the last commit and when I change
```xml
<sch:rule context="//text()[ancestor::*[@xml:lang = 'en-US']]">
```

to 

```xml
<sch:rule context="//*[@xml:lang = 'en-US']">
```

It seems to work out. Just select all nodes that have the "en-US" language, and then calling `text()` on them recursively assembles the text node.

Please keep in mind, that the selection of these nodes may overlap with other nodes, so it should be clearig in its own pattern.

hth

**Reply by pradyumanaggarwal** [reply]
Thanks for this @phax, just want to confirm one more thing, XPATH expressions in context attribute is supported, right ?

**Reply by phax** [reply]
Yes absolutely - as e.g. `//*[@xml:lang = 'en-US']` is already an XPath expression. If you use XSLT 2.0 as the query binding, XPath 2.0 is allowed. When you use the XSLT binding, you can even use XSLT expressions in the context.

**Reply by pradyumanaggarwal** [reply]
Problem occurs when i try to use `context = "//text()"`. What it could be, like is it not able to specify one node or am i missing anything ?

**Reply by phax** [reply]
I think that is an XPath specialty. As you can generate `text()` from any node. I am not that deep into XPath to explain it in all detail, but non-element nodes are usually tricky. So try to always base your selectors on elements (like `*` ;-) ) and take it from there. E.g. be more precise in single asserts.

**Reply by pradyumanaggarwal** [reply]
Thanks, it helps a lot.
Do we have any documentation for the limitations in XPATHS or anything like that ?
@phax

**Reply by phax** [reply]
Well, there is the XPath 2.0 Specification from W3C....

https://www.w3.org/TR/xpath20/

**Reply by pradyumanaggarwal** [reply]
Does this means we don’t have support for XPATH 3.0?

**Reply by phax** [reply]
That depends on the XPath processor you are using. Saxon it is by default when using ph-schematron. I *think* it supports XPath 3.0 but please crosscheck on their side....
